### DISCUSSION TITLE
Transactions in Record Request Event Hooks

### DISCUSSION THREAD

**Discussion by hlanderdev** [pocketbase/pocketbase#3146] [Q&A] [answered]
Hi, I have a couple of scenarios where I would like to run code within the same transaction as the record being processed in a request event hooks (i.e. OnRecordBeforeCreateRequest).

The first scenario, is needing to automatically create/update additional records to reflect the record being created/updated via the API. For example, a user creates an "organizations" record via a Pocketbase API call and needs to automatically have a "roles" record created to mark them as an admin in the organization. If either the "organizations" or "roles" fails to save, we would like to fail the entire transaction for data integrity.

The second scenario is that we have a caching feature for our project using an external API. For example, when the user creates an "events" record, we generate a static web page by using a Git web API. As far as I know the API we are using does not have a transaction feature, so ideally we make this call at the very end, after the record has been created in Pocketbase but the transaction has not been committed yet. That way if the API call fails we do not end up with any "events" records with "published" set to true despite the page not actually existing.

In both of these scenarios we need the HTTP request context, but a transaction dao does not seem to be available in the event structs for those event hooks (i.e. core.RecordCreateEvent). Is there any way to achieve this functionality currently that I missed? If not is it possible to add the transaction dao to the request hook event structs?

**Comment by ganigeorgiev** [comment, marked as answer]
We currently don't wrap by default the create/update request record persistence in a explicit transaction in order to minimize the db locking times (_especially if the transaction has mixed read/write operations_). It is a trade-off between consistency/correctness vs performance and someone is bound to be unhappy.

Unfortunately I'm not sure what to suggest at the moment. I've added it in the [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1) to think a little more on it for v0.18.0/v0.19.0.

**Reply by hlanderdev** [reply]
Ok, no problem! Thanks for the quick reply!

**Comment by mez** [comment]
Any plans for this feature yet?

**Reply by ganigeorgiev** [reply]
It exists in the roadmap but there are no ETAs.

**Comment by nexovec** [comment]
Are transactions possible with pocketbase?

**Reply by starredev** [reply]
> Are transactions possible with pocketbase?

https://pocketbase.io/docs/js-database/#transaction

**Reply by ganigeorgiev** [reply]
@nexovec You are commenting to a very old discussion and it is not clear what do you mean by just _"Are transactions possible"_. As others already linked, db transactions are supported since the first release but this discussion is for the **request event hooks** and support for wrapping the entire hooks chain in a transaction was added in v0.23+.

In any case, please don't hijack old discussions and instead open a new one clearly elaborating on your use case and what you are trying to do (see also [XY problem](https://xyproblem.info/)).

**Comment by ganigeorgiev** [comment]
_Just in case someone stumble here - option to wrap the request hooks chain in a transaction was added in PocketBase v0.23+. For example:_

- Go:
    ```go
    app.OnRecordCreateRequest("users").BindFunc(func(e *core.RecordRequestEvent) error {
        originalApp := e.App
        txErr := e.App.RunInTransaction(func(txApp core.App) error {
            e.App = txApp

            // do something before the create...

            // execute the create
            err := e.Next()
            if err != nil {
                return err
            }

            // do something after the create...

            return nil
        })
        e.App = originalApp

        return txErr
    })
    ```

- JS:
    ```js
    onRecordCreateRequest((e) => {
        let originalApp = e.app
        try {
            e.app.runInTransaction((txApp) => {
                e.app = txApp

                // do something before the create...

                // execute the create
                e.next()

                // do something after the create...
            })
        } finally {
            e.app = originalApp
        }
    }, "users")
    ```

**Reply by pchasco** [reply]
Just for clarification on this, following this pattern if I create a record for collection A, then in the on RecordAfterCreateSuccess I create a record for collection B, if B fails and rolls back the transaction, the A record would not be committed? And then if that's the case, would this continue recursively for all hooks that are run as a result of that very first A record that was created? I'm wondering if the onRecordCreateRequest handler above would need to be modified so that it is aware if it is being reentered with an app that is already running in the context of a transaction.

```
app.OnRecordCreateRequest().BindFunc(func(e *core.RecordRequestEvent) error {
  if e.App.IsTransactional() {
    return e.Next()
  }
  
  originalApp := e.App  
  txErr := e.App.RunInTransaction(func(txApp core.App) error {
    e.App = txApp    
    return e.Next()
  })
  e.App = originalApp
  
  return txErr
})
```

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand what you are asking, but if you want your second record B create to run as part of your transaction AND after the record A create, then you can use the regular [`OnRecordCreate`](https://pocketbase.io/docs/go-event-hooks/#onrecordcreate) hook instead of `OnRecordAfterCreateSuccess` _(or alternatively move the record B create logic inside the started transaction and after the `e.Next` call)_. For example:

```go
// OnRecordCreateRequest should also work and could be technically the more correct hook here
// because it ensures that the side-effect will run only as part of the request operation
// but it really depends what actually is the end goal here
app.OnRecordCreate("A").BindFunc(func(e *core.RecordEvent) error {
    if err := e.Next(); err != nil {
        return err
    }

    // create record B after successfull record A creation
    // (note if record A create was triggered by OnRecordCreateRequest and B fails then A would be reverted because of the OnRecordCreateRequest transaction)
    collectionB, _ := e.App.FindCachedCollectionByNameOrId("B")
    record := core.NewRecord(collectionB)
    return e.App.Save(record)
})
```

In general the hooks [`OnRecordAfterCreateSuccess`](https://pocketbase.io/docs/go-event-hooks/#onrecordaftercreatesuccess)/[`OnRecordAfterCreateError`](https://pocketbase.io/docs/go-event-hooks/#onrecordaftercreateerror) are special in the sense that they are called AFTER the transaction is **"done"** (committed or failed). This means that if you decide to create a new record in one of those 2 hooks then it is expected the operation to run **outside** of the defined transaction in your `OnRecordCreateRequest`.

**Reply by pchasco** [reply]
OK, I think you hit on what I was looking for. Essentially, I want to insert several records into different collections when a client attempts to create record of type "A." I want that operation to be atomic. It looks like if I use OnRecordCreate hook I will be covered. Thanks!

**Reply by ganigeorgiev** [reply]
> I want that operation to be atomic. It looks like if I use OnRecordCreate hook I will be covered. 

Yes, that should suffice.

_As a side-note keep in mind that the `OnRecordCreate` hook fires also for manually/programmatic created records (aka. when calling `app.Save(record)`), meaning that if your transaction wrapping happens in the `OnRecordCreateRequest` hook it is possible that there could be a situation where it won't be atomic because there will be no transaction since the request hook won't be triggered. In that case, you can either move the transaction wrapping inside `OnRecordCreate` OR if you want the transaction wrapping to happen only for the API request, replace `OnRecordCreate` with `OnRecordCreateRequest`. Again, it really depends on the end goal and desired workflow._
